File: Week_09/917/reverse_only_letters.go
package main

func isLetter(c byte) bool {
	if c >= 'A' && c <= 'Z' || c >= 'a' && c <= 'z' {
		return true
	}
	return false
}

func reverseOnlyLetters(S string) string {
	bytes := []byte(S)
	left, right := 0, len(bytes)-1
	for left < right {
		if !isLetter(bytes[left]) {
			left++
		} else if !isLetter(bytes[right]) {
			right--
		} else {
			bytes[left], bytes[right] = bytes[right], bytes[left]
			left++
			right--
		}
	}
	return string(bytes)
}

File: Week_09/917/reverse_only_letters.py
class Solution:
    def reverseOnlyLetters(self, S: str) -> str:
        left, right = 0, len(S) - 1
        chars = list(S)
        while left < right:
            if not 'A' <= chars[left] <= 'Z' and not 'a' <= chars[left] <= 'z':
                left += 1
            elif not 'A' <= chars[right] <= 'Z' and not 'a' <= chars[right] <= 'z':
                right -= 1
            else:
                chars[left], chars[right] = chars[right], chars[left]
                left += 1
                right -= 1
        return "".join(chars)
